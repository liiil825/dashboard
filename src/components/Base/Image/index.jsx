import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { startsWith } from 'lodash';

import styles from './index.scss';

export default class Image extends React.Component {
  static propTypes = {
    className: PropTypes.string,
    iconLetter: PropTypes.string,
    iconSize: PropTypes.number,
    src: PropTypes.string
  };

  static defaultProps = {
    iconSize: 24,
    src: '',
    iconLetter: ''
  };

  state = {
    failed: false
  };

  componentDidMount() {
    this.img.onerror = () => {
      this.setState({ failed: true });
    };
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      nextProps.src !== this.props.src
      || Boolean(nextState.failed)
      || Boolean(nextProps.iconLetter)
    );
  }

  componentWillUnmount() {
    if (this.img && this.img.onerror) {
      this.img.onerror = null;
    }
  }

  render() {
    const {
      src, iconLetter, className, iconSize, ...rest
    } = this.props;
    const { failed } = this.state;
    const imgStr = src && startsWith(src, 'att-') ? `/attachments/${src}` : src;

    if (failed) {
      const style = {
        width: `${iconSize}px`,
        height: `${iconSize}px`
      };
      const letter = iconLetter.substr(0, 1).toLocaleUpperCase();

      if (letter) {
        style.fontSize = `${iconSize / 2}px`;
        style.padding = `${iconSize / 4 - 1}px`;

        return (
          <span
            className={classnames(styles.letter, className)}
            style={style}
            {...rest}
          >
            {letter}
          </span>
        );
      }

      return (
        <img
          src="/none.svg"
          data-origin-url={imgStr}
          style={style}
          className={classnames(styles.img, className)}
          {...rest}
        />
      );
    }

    return (
      <img
        src={imgStr}
        data-origin-url={imgStr}
        className={classnames(styles.img, className)}
        ref={c => {
          this.img = c;
        }}
        {...rest}
      />
    );
  }
}
