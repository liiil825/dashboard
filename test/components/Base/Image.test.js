import Image from 'components/Base/Image';

describe('Base/Iamge', () => {
  const LOAD_FAILURE_SRC = 'LOAD_FAILURE_SRC';
  const LOAD_SUCCESS_SRC = 'https://cc_webchat.yunify.com/images/icon.png';

  it('basic render', () => {
    const wrapper = render(<Image src={LOAD_SUCCESS_SRC} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('render base64', () => {
    const wrapper = render(<Image src={LOAD_SUCCESS_SRC} isBase64Str />);
    expect(wrapper).toMatchSnapshot();
  });

  it('on error', () => {
    const wrapper = mount(
      <Image src={LOAD_FAILURE_SRC} iconLetter={LOAD_FAILURE_SRC} isBase64Str />
    );
    const img = wrapper.instance().img;
    img.onerror(new Error('mocked error'));
    expect(wrapper.state().failed).toEqual(true);

    wrapper.unmount();
  });
});
