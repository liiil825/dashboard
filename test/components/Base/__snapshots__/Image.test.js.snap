// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base/Iamge basic render 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "img",
      "data-origin-url": "https://cc_webchat.yunify.com/images/icon.png",
      "src": "https://cc_webchat.yunify.com/images/icon.png",
    },
    "children": Array [],
    "name": "img",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "data-origin-url": undefined,
      "src": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "data-origin-url": undefined,
      "src": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Base/Iamge render base64 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "img",
      "data-origin-url": "data:image/png;data:image/svg;data:image/jpg;base64,https://cc_webchat.yunify.com/images/icon.png",
      "src": "data:image/png;data:image/svg;data:image/jpg;base64,https://cc_webchat.yunify.com/images/icon.png",
    },
    "children": Array [],
    "name": "img",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "data-origin-url": undefined,
      "src": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "data-origin-url": undefined,
      "src": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
